README.md
# infini
====

# Description
デバイス多台数検証用スクリプトとデータマッチングスクリプト

# Contents
- デバイス多台数検証用スクリプト（main2.py, params.py）
- データマッチングスクリプト（matching.py）

# Usage
## デバイス多台数検証用スクリプト
指定台数の仮想デバイスを起動し、MQTT を使用してサーバーにデータを送信します。

### 設定
`params.py` で行います。

- `USERNAME`: MQTT クライアントのユーザーネーム
- `PASSWORD`: MQTT クライアントのパスワード
- `HOST`: MQTT ブローカーの ホスト名または IP アドレス

- `N_DEVICE`: 起動する仮想デバイスの数

- `BEACON_PERIOD`: ビーコンデータを送出する間隔（秒）
- `ENV_PERIOD`:  内部環境情報を送出する間隔（秒）
- `DIGITAL_SENSOR_PERIOD`: デジタルセンサーデータを送出する間隔（秒）
- `DIGITAL_COUNTER_PERIODS`: デジタルカウンターデータを送出する間隔（秒）

### 起動
`$ python main2.py`

## データマッチングスクリプト
サーバーからダウンロードしたセンサーデータ（以下、サーバーデータ）と、仮想デバイスからデータ送出時にローカルに保存したセンサーデータ（以下、ローカルデータ）のマッチングを行います。

### 設定
`matching.py` 内で行います。

- `device_id`: 何番のデバイスのデータをマッチングするかを指定
- `sensor_names`: マッチングの対象とするセンサーの名前、内部環境情報（`Envinfo`）、アナログセンサー（`Analog_sensors`）、デジタルセンサー（`Digital_sensors`）、デジタルカウンター（`Digital_counters`）、ビーコン（`Beacon`）
- `time_stamp`: マッチングを行うローカルデータのタイムスタンプ
- `begin`: この時刻からのデータをマッチングの対象とする
- `end`: この時刻までのデータをマッチングの対象とする

### 起動
`$ python matching.py`

### 出力
センサー名に対してマッチングの結果（データに不一致や欠落がなければ `OK`、あれば `NG`）が出力される

# Install
以下のコマンドで任意のディレクトリに clone してください。  
`$ git clone git@github.com:matsu490/infini.git`  
または任意のディレクトリに ZIP ファイルをダウンロードして解凍してください。

# Requirement
- Python 2.7.x
- Pandas

# Author
[matsu490](https://github.com/matsu490)


CHANGELOG.md
# Change Log

## [v0.1.3](https://github.com/matsu490/infini/tree/v0.1.3) (2018-04-11)
**Implemented enhancements:**
- ログファイルをエクセルで開いている時に生じる、書き込み不可エラーへの対処をした（チケット #2464 参照）
- マッチングの際、judge_table に False が含まれていても OK 判定を返すバグの修正をした（チケット #2464 参照）

## [v0.1.2](https://github.com/matsu490/infini/tree/v0.1.2) (2018-01-17)
**Implemented enhancements:**
- typo (Envinfo -> EnvInfo)
- 内部環境情報のカラムラベルの順序を変更した（チケット #2458 参照）
- ANALOG_SENSOR_PERIOD を一時的に追加した（アナログセンサーグループの実装とともに消す予定）

## [v0.1.1](https://github.com/matsu490/infini/tree/v0.1.1) (2018-01-16)
**Implemented enhancements:**
- タイムスタンプの月、日を 0 埋めして二桁表示にした

## [v0.1.0](https://github.com/matsu490/infini/tree/v0.1.0) (2017-12-27)
**Implemented enhancements:**
- 初めてのリリース
- デバイス多台数検証用スクリプト
- データマッチングスクリプト

## ToDo
- [ ] データマッチングスクリプト、読み込みデータの不具合による例外処理を追加する
- [x] アナログセンサーグループの実装


main.py
# -*- coding: utf-8 -*-
# vim: set fileencoding=utf-8 :
# vim: set foldmethod=marker commentstring=\ \ #\ %s :
#
# Author:    Taishi Matsumura
# Created:   2017-10-31
#
# Copyright (C) 2017 Taishi Matsumura
#
import sys
import time
import paho.mqtt.publish as publish
import threading
import random
import csv
import os
import os.path
import datetime
from params import *


class Device(threading.Thread):
    def __init__(self, username, password, device_id, message, host,
            period=15, port=1883,
            analog=False, csv=False, save_dir='./MG2_send_data/'):
        super(Device, self).__init__()
        self.period = period
        self.host = host
        self.port = port
        self.username = username
        self.password = password
        self.device_id = device_id
        self.topic = '%s/%s' % (password, device_id)
        self.message = message
        self.analog = analog
        self.csv = csv
        self.save_dir = save_dir
        self.init_csv()
        self.setDaemon(True)

    def init_csv(self):
        if self.csv:
            if not os.path.exists(self.save_dir):
                os.mkdir(self.save_dir)
            with open('{0}{1}.csv'.format(self.save_dir, self.device_id), 'w') as f:
                writer = csv.writer(f, lineterminator='\n')
                if self.analog:
                    header = ['tm','a1','a2','a3','a4','a5','a6','a7','a8','msg']
                else:
                    header = ['tm','msg']
                writer.writerow(header)

    def run(self):
        while True:
            tm = time.time()
            if self.analog:
                data = [tm, '300.00', '2.00', '3.00', '4.00', '5.00', '6.00', '7.00', '8.00', self.message]
                payload = '{{"tm":"{0}","a1":{1},"a2":{2},"a3":{3},"a4":{4},"a5":{5},"a6":{6},"a7":{7},"a8":{8},"msg":"{9}"}}'.format(*data)
            else:
                data = [tm, self.message]
                payload = '{{"tm":"{0}","msg":"{1}"}}'.format(*data)
            publish.single(topic='%s/%s' % (self.password, self.device_id),
                    payload=payload,
                    hostname=self.host,
                    auth={'username': self.username, 'password': self.password})
            if self.csv:
                d = datetime.datetime.fromtimestamp(tm)
                data[0] = '{0}-{1}-{2} {3:02d}:{4:02d}:{5:02d}.{6}'.format(d.year, d.month, d.day, d.hour, d.minute, d.second, str(round(1e-6 * d.microsecond, 2))[2:])
                with open('{0}{1}.csv'.format(self.save_dir, self.device_id), 'a') as f:
                    writer = csv.writer(f, lineterminator=',\n')
                    writer.writerow(data)
            print tm, self.message
            time.sleep(self.period)


class Devices(object):
    def __init__(self, n, period, host, analog=False, csv=False):
        self.n = n
        self.devices = {}
        for i in xrange(1, n+1):
            line = "self.devices['d{0:04d}'] = Device('MG00017', '4w2P47Ph', 'IFT_ML1-YONEZAWA{1:04d}', '検証機{2:04d}', host, period={3}, analog={4}, csv={5})".format(i, i, i, period, analog, csv)
            exec(line)

    def start(self):
        for i in xrange(1, self.n+1):
            time.sleep(1 * random.random())
            line = "self.devices['d{0:04d}'].start()".format(i)
            exec(line)

if __name__ == '__main__':
    ds = Devices(n=100, period=15, host=host, analog=True, csv=True)
    ds.start()

    while True:
        c = sys.stdin.read(1)
        if c == 'e':
            sys.exit()


matching.py
# -*- coding: utf-8 -*-
# vim: set fileencoding=utf-8 :
# vim: set foldmethod=marker commentstring=\ \ #\ %s :
#
# Author:    Taishi Matsumura
# Created:   2017-11-16
#
# Copyright (C) 2017 Taishi Matsumura
#
import pandas as pd

device_id = 1
sensor_names = ['EnvInfo', 'Analog_group1', 'Digital_sensors', 'Digital_counters', 'Beacon']
time_stamp = '20180117124526'


# server_data: load the data downloaded from the server
headers = ['time'] + ['d{}'.format(i+1) for i in xrange(8)] + ['a{}'.format(i+1) for i in xrange(8)] + ['eiTemp', 'eiHumi', 'eiLPrs', 'seaPrs', 'beacon', 'message']
server_data_raw = pd.read_csv('./receivedata.csv', header=0, names=headers).set_index('time')
begin = server_data_raw.index[0]
end = server_data_raw.index[-1]

for sensor_name in sensor_names:
    # local_data: load the data stored by python in local
    file_path = './Logs/IFT_ML1-YONEZAWA{:04d}/{}_{}.csv'.format(device_id, sensor_name, time_stamp)
    tmp_data = pd.read_csv(file_path).set_index('time')
    tmp_data = tmp_data.loc[begin:end, :]
    tmp_data = tmp_data.iloc[:, 0:-1]
    local_data = {sensor_name: tmp_data}

    # extract data of the sensor from the server_data
    server_data = {sensor_name: server_data_raw.loc[begin:end, tmp_data.columns].dropna(how='all', axis=0)}

    df = pd.concat([local_data[sensor_name], server_data[sensor_name]], axis=1)
    judge_table = local_data[sensor_name] == server_data[sensor_name]
    result = all(judge_table.all())
    print judge_table

    okng = 'OK'*result + 'NG'*(not result)
    print '{}: {}'.format(sensor_name, okng)


scraping.py
# -*- coding: utf-8 -*-
# vim: set fileencoding=utf-8 :
# vim: set foldmethod=marker commentstring=\ \ #\ %s :
#
# Author:    Taishi Matsumura
# Created:   2017-11-10
#
# Copyright (C) 2017 Taishi Matsumura
#
import bs4
import requests
import selenium.webdriver
import time
import csv
import os.path
from params import *

driver = selenium.webdriver.Chrome('./chromedriver.exe')
# driver = selenium.webdriver.Firefox('.')
driver.get(url)
driver.maximize_window()
uid = driver.find_element_by_name('login_id')
pw = driver.find_element_by_name('password')
uid.send_keys(user_name)
pw.send_keys(password)
# Click the login button
driver.find_element_by_xpath('//*[@id="LoginPanel"]/form/input[3]').click()

driver.get('https://m-linker.infinitec.co.jp/MobileGates2/ReceivedataView')
time.sleep(5)
# Expand the tree for devices
driver.find_element_by_xpath('//*[@id="treeView"]/ul/li[2]/span[2]').click()

# Select a device
driver.find_element_by_xpath('//*[@id="treeView"]/ul/li[4]').click()

# Select the list tab
driver.find_element_by_xpath('//*[@id="listfieldseltab"]').click()

# Check the box for message
driver.find_element_by_xpath('//*[@id="list_checkbox_msg"]/td[1]/input').click()

# Output a .csv file of the table data
elements = driver.find_element_by_id('datePeriod')
select_elements = selenium.webdriver.support.ui.Select(elements)
select_elements.select_by_index(8)
# Specify the period to show
driver.find_element_by_id('startDateValue').clear()
driver.find_element_by_id('endDateValue').clear()
driver.find_element_by_id('startDateValue').send_keys('2017-11-14 10:30')
driver.find_element_by_id('endDateValue').send_keys('2017-11-14 10:35')
driver.find_element_by_name('recievedata').click()
# Download the .csv file
time.sleep(5)
driver.find_element_by_xpath('//*[@id="dataExport"]').click()


main2.py
# -*- coding: utf-8 -*-
# vim: set fileencoding=utf-8 :
# vim: set foldmethod=marker commentstring=\ \ #\ %s :
#
# Author:    Taishi Matsumura
# Created:   2017-10-31
#
# Copyright (C) 2017 Taishi Matsumura
#
import os
import sys
import csv
import time
import datetime
import paho.mqtt.publish as publish
import threading
import numpy as np
from params import *


class Sensor(object):
    def __init__(self, name, device_id):
        self.sensor_name = name
        self.device_id = device_id
        self.header = []
        self.logfile_path = ''
        self.temp_logfile_path = ''
        self.io_error = False

    def _init_logfile(self):
        self._make_logdir()
        self.logfile_path = './Logs/{}/{}_{}.csv'.format(self.device_id, self.sensor_name, self._timestamp())
        self.temp_logfile_path = os.path.splitext(self.logfile_path)[0] + '.tmp'
        self._make_logfile(self.logfile_path, self.header + ['is_err'])

    def _timestamp(self):
        d = datetime.datetime.fromtimestamp(time.time())
        return '{0}{1:02d}{2:02d}{3:02d}{4:02d}{5:02d}'.format(d.year, d.month, d.day, d.hour, d.minute, d.second)

    def _make_logdir(self):
        try:
            os.mkdir('./Logs/{}'.format(self.device_id))
        except:
            pass

    def _logging(self, is_err):
        d = datetime.datetime.fromtimestamp(self.data[0])
        dtime = '{0}-{1:02d}-{2:02d} {3:02d}:{4:02d}:{5:02d}.{6}'.format(d.year, d.month, d.day, d.hour, d.minute, d.second, str(round(1e-6 * d.microsecond, 2))[2:])
        line = [dtime] + self.data[1:] + [is_err]
        try:
            self._write_log(self.logfile_path, line)
        except IOError:
            print 'The log was not written because the logfile is open by, maybe, MS Excel.'
            if self.io_error:
                pass
            # If the IOError didn't happen previously, make a new templogfile.
            else:
                self._make_logfile(self.temp_logfile_path, self.header + ['is_err'])
            self._write_log(self.temp_logfile_path, line)
            self.io_error = True
        else:
            # If the IOError happend previously, merge and remove the templogfile.
            if self.io_error:
                self._merge_templogfile()
                os.remove(self.temp_logfile_path)
            self.io_error = False

    def _merge_templogfile(self):
        with open(self.logfile_path, 'r') as f:
            log_file = f.readlines()
        with open(self.temp_logfile_path, 'r') as f:
            temp_log = f.readlines()
        log_file[-1:0] = temp_log[1:]
        with open(self.logfile_path, 'w') as f:
            f.writelines(log_file)

    def _make_logfile(self, file_path, line):
        with open(file_path, 'w') as f:
            writer = csv.writer(f, lineterminator='\n')
            writer.writerow(line)

    def _write_log(self, file_path, line):
        with open(file_path, 'a') as f:
            writer = csv.writer(f, lineterminator='\n')
            writer.writerow(line)

    def run(self):
        time.sleep(5 * np.random.rand())
        while True:
            self._make_sensor_data()
            self._send_sensor_data()
            time.sleep(self.period)

    def _make_sensor_data(self):
        pass

    def _send_sensor_data(self):
        try:
            publish.single(topic='{}/{}'.format(self.password, self.device_id),
                    payload=self.payload, hostname=self.host,
                    auth={'username': self.username, 'password': self.password})
            self._logging(0)
            print '{}: {}\n'.format(self.sensor_name, self.payload)
        except:
            self._logging(1)
            print '{}: The payload was not send.'.format(self.sensor_name)


class Beacon(Sensor, threading.Thread):
    def __init__(self, name, username, password, host, device_id, period, beacon):
        super(Beacon, self).__init__(name, device_id)
        super(Sensor, self).__init__()
        self.username = username
        self.password = password
        self.host = host
        self.device_id = device_id
        self.period = period
        self.beacon = beacon
        self.header = ['time', 'beacon']
        self._init_logfile()
        self.setDaemon(True)

    def _make_sensor_data(self):
        tm = round(time.time(), 2)
        self.data = [tm, self.beacon]
        self.payload = '{{"tm":"{0}","Beacon":"{1}"}}'.format(*self.data)


class EnvironmentalInformation(Sensor, threading.Thread):
    def __init__(self, name, username, password, host, device_id, period):
        super(EnvironmentalInformation, self).__init__(name, device_id)
        super(Sensor, self).__init__()
        self.username = username
        self.password = password
        self.host = host
        self.device_id = device_id
        self.period = period
        self.t = 0
        self.dt = 0.1
        self.header = ['time', 'eiTemp', 'eiHumi', 'eiLPrs', 'seaPrs']
        self._init_logfile()
        self.setDaemon(True)

    def _make_sensor_data(self):
        tm = round(time.time(), 2)
        eiTemp = round(np.random.rand() + 20   +  10*np.sin(2*np.pi*self.t), 2)
        eiHumi = round(np.random.rand() + 50   +  30*np.sin(2*np.pi*self.t), 2)
        eiLPrs = round(np.random.rand() + 900  + 100*np.sin(2*np.pi*self.t), 2)
        seaPrs = round(np.random.rand() + 1000 + 100*np.sin(2*np.pi*self.t), 2)
        self.t += self.dt
        self.data = [tm, eiTemp, eiHumi, eiLPrs, seaPrs]
        self.payload = '{{"tm":"{0}","eiTemp":{1},"eiHumi":{2},"eiLPrs":{3},"seaPrs":{4}}}'.format(*self.data)


class DigitalSensors(Sensor, threading.Thread):
    def __init__(self, name, username, password, host, device_id, port_ids, period):
        super(DigitalSensors, self).__init__(name, device_id)
        super(Sensor, self).__init__()
        self.username = username
        self.password = password
        self.host = host
        self.device_id = device_id
        self.port_ids = port_ids
        self.period = period
        self.header = ['time'] + ['d{}'.format(port_id) for port_id in port_ids]
        self._init_logfile()
        self.setDaemon(True)

    def _make_sensor_data(self):
        tm = round(time.time(), 2)
        self.data = [tm] + [np.random.randint(0, 2) for _ in self.port_ids]
        self.payload_ = ['"tm":"{0}"'.format(self.data[0])]
        for port_id, randint in zip(self.port_ids, self.data[1:]):
            self.payload_.append('"d{0}":{1}'.format(port_id, randint))
        self.payload = '{{{0}}}'.format(','.join(self.payload_))


class DigitalCounters(object):
    def __init__(self, username, password, host, device_id, port_ids, periods):
        self.port_ids = port_ids
        self.periods = periods
        self.counters = {}
        for i, (port_id, period) in enumerate(zip(port_ids, periods)):
            self.counters[i] = DigitalCounter('Digital_counters', username, password, host, device_id, port_id, period)

    def run(self):
        for counter in self.counters.values():
            counter.start()


class DigitalCounter(Sensor, threading.Thread):
    def __init__(self, name, username, password, host, device_id, port_id, period):
        super(DigitalCounter, self).__init__(name, device_id)
        super(Sensor, self).__init__()
        self.username = username
        self.password = password
        self.host = host
        self.device_id = device_id
        self.port_id = port_id
        self.period = period
        self.header = ['time', 'd{}'.format(port_id)]
        self._init_logfile()
        self.setDaemon(True)

    def _make_sensor_data(self):
        tm = round(time.time(), 2)
        self.data = [tm, np.random.randint(0, 50)]
        self.payload = '{{"tm":"{0}","d{1}":{2}}}'.format(self.data[0], self.port_id, self.data[1])


class DigitalElements(object):
    def __init__(self, username, password, host, device_id, global_period, periods=8*[0]):
        sensor_ids = [i+1 for i, p in enumerate(periods) if p==0]
        counter_ids = list(set(xrange(1, 9)) - set(sensor_ids))
        counter_periods = [p for p in periods if p!=0]
        self.sensors = DigitalSensors('Digital_sensors', username, password, host, device_id, sensor_ids, global_period)
        self.counters = DigitalCounters(username, password, host, device_id, counter_ids, counter_periods)

    def run(self):
        self.sensors.start()
        self.counters.run()


class AnalogSensors(Sensor, threading.Thread):
    def __init__(self, name, username, password, host, device_id, period, check_box):
        super(AnalogSensors, self).__init__(name, device_id)
        super(Sensor, self).__init__()
        self.username = username
        self.password = password
        self.host = host
        self.device_id = device_id
        self.period = period
        self.check_box = check_box
        self.port_ids = [i+1 for i in xrange(len(check_box)) if check_box[i]==1]
        self.header = ['time'] + ['a{}'.format(port_id) for port_id in self.port_ids]
        self.t = 0
        self.dt = 0.1
        self._init_logfile()
        self.setDaemon(True)

    def _make_sensor_data(self):
        tm = round(time.time(), 2)
        waves = [round(np.random.rand() + 10 + 10*np.sin(2*np.pi*self.t), 2),
                 round(np.random.rand() + 20 + 20*np.sin(2*np.pi*self.t), 2),
                 round(np.random.rand() + 30 + 30*np.sin(2*np.pi*self.t), 2),
                 round(np.random.rand() + 40 + 40*np.sin(2*np.pi*self.t), 2),
                 round(np.random.rand() + 10 + 10*np.cos(2*np.pi*self.t), 2),
                 round(np.random.rand() + 20 + 20*np.cos(2*np.pi*self.t), 2),
                 round(np.random.rand() + 30 + 30*np.cos(2*np.pi*self.t), 2),
                 round(np.random.rand() + 40 + 40*np.cos(2*np.pi*self.t), 2)]
        self.data = [tm] + [waves[i-1] for i in self.port_ids]
        self.payload_ = ['"tm":"{0}"'.format(self.data[0])]
        for port_id, wave in zip(self.port_ids, self.data[1:]):
            self.payload_.append('"a{0}":{1}'.format(port_id, wave))
        self.payload = '{{{0}}}'.format(','.join(self.payload_))
        self.t += self.dt


class Device(object):
    def __init__(self, username, password, host, device_id, p_beacon=BEACON_PERIOD, p_env=ENV_PERIOD, p_digi=DIGITAL_SENSOR_PERIOD):
        self.username = username
        self.password = password
        self.host = host
        self.device_id = device_id
        self.beacon = Beacon('Beacon', username, password, host, device_id, p_beacon, beacon=device_id)
        self.envinfo = EnvironmentalInformation('EnvInfo', username, password, host, device_id, p_env)
        self.digital_elems = DigitalElements(username, password, host, device_id, global_period=p_digi, periods=DIGITAL_COUNTER_PERIODS)
        self.anagroup1 = AnalogSensors('Analog_group1', username, password, host, device_id, ANALOG_GROUP1_PERIOD, ANALOG_GROUP1)
        self.anagroup2 = AnalogSensors('Analog_group2', username, password, host, device_id, ANALOG_GROUP2_PERIOD, ANALOG_GROUP2)
        self.anagroup3 = AnalogSensors('Analog_group3', username, password, host, device_id, ANALOG_GROUP3_PERIOD, ANALOG_GROUP3)
        self.anagroup4 = AnalogSensors('Analog_group4', username, password, host, device_id, ANALOG_GROUP4_PERIOD, ANALOG_GROUP4)

    def switch_on(self):
        self.beacon.start()
        self.envinfo.start()
        self.digital_elems.run()
        self.anagroup1.start()
        self.anagroup2.start()
        self.anagroup3.start()
        self.anagroup4.start()


class Devices(object):
    def __init__(self, device_name, n, username, password, host):
        self.device_name = device_name
        self.n = n
        self.username = username
        self.password = password
        self.host = host
        self.devices = {}
        self._init_devices()

    def _init_devices(self):
        if self.n == 1:
            self.devices[1] = Device(self.username, self.password, self.host, self.device_name)
        elif self.n > 1:
            for i in xrange(1, self.n+1):
                device_id = '{0}{1}'.format(self.device_name, '{:04d}'.format(i))
                self.devices[i] = Device(self.username, self.password, self.host, device_id)
        else:
            raise DataRangeError()

    def switch_on(self):
        for i in xrange(1, self.n+1):
            self.devices[i].switch_on()


class DataRangeError(Exception):
    def __str__(self):
        return 'The value is out of range.'


if __name__ == '__main__':
    devices = Devices(DEVICE_NAME, N_DEVICE, USERNAME, PASSWORD, HOST)
    devices.switch_on()

    while True:
        c = sys.stdin.read(1)
        if c == 'e':
            sys.exit()
